### ISSUE TITLE
Worker-local queues

### ISSUE THREAD

**Issue by magik6k** [filecoin-project/sector-storage#17]
Each worker should have a small queue of tasks it could be prefetching while others are executing

Currently:
* Workers declare available resources to the scheduler
* Scheduler assigns tasks to workers from main queue, based on those resources, and marks them as reserved until the task is complete
* If the worker needs to fetch sector data, it won't be able to do any work while that happens, meaning that it's resources won't be fully used

Idea v0:
* Add small queues to workers (keep main queue in the scheduler)
* Queue depth based on resources, let's say `2 * min(cpu, mem, disk)` (tunable by the user) 
* Assigning work to small queues must be done in some way which avoids starving bigger tasks with smaller tasks, and filling all cache disks
  * One way to do that would be for the main scheduler to always assign tasks later in the pipeline (commit2 instead precommit1 for example) to workers which can execute them, in case there aren't enough resources for the preferred task, wait for them to become available instead of throwing smaller tasks at the worker.

Per worker queues should also enable 'multi-master' workers, and if the interfaces turn out nicely, multi-level scheduling which would simplify things like taking data locality in racks/datacenters into account

**Comment by icorderi** [comment]
A local work queue has its own set of pros/cons. 
- What is the main motivation for this?

Starvation is an issue only if the work ordering strategy allows for it. We have simple solutions available to use here (for instance take age into account).   

> main scheduler to always assign tasks later in the pipeline (commit2 instead precommit1 for example)
This is good, it touches on pipeline optimization. 

I would first consider **why** we need edge queues:
- is the cost of fetching the next item too high compared to cost of running it? 
- Is it because the scheduler can disappear and we want to continue working? 
- Are we capable of running more cost intensive computations to optimize a local execution plan?
- Do we assume a constant filled work pipe? 
- Are we going to support work stealing? 

Separately, I would consider pipeline optimization strategies like the one you mentioned (task commit2 instead precommit1) 
- what does it win us? 
- What are others like it we can use  

> Per worker queues should also enable 'multi-master' workers
This can be also achieved by keeping the work queue backed by consensus. 
Why might want to consider pros/cons of each. 

> multi-level scheduling which would simplify things like taking data locality in racks/datacenters into account
Yes, we need to consider sector affinity during the scheduling 
- what is the cost of moving the next stage to another node vs doing it on the same node? 
I'm not sure local queues necessarily solves for this. You are going to face the same problem that if you don't have them.

**Comment by magik6k** [comment]
> I would first consider why we need edge queues
> * is the cost of fetching the next item too high compared to cost of running it?

Yes, sector cache for precommit2 is in the 350-400G range last time I checked, and moving that over 1Gbps (which if we can we should support as best as we can just based on how much cheaper it is) takes ~1h.

Precommit2 compute takes 51min on a 32 core AMD Threadripper, which means that most of the time is spend just waiting. And as sealing becomes faster, this will become an even bigger issue.

> Is it because the scheduler can disappear and we want to continue working?

That's a separate issue, but we could use that as a part of addressing this

> Are we capable of running more cost intensive computations to optimize a local execution plan?

I'm assuming very small edge queues (2-3 tasks), I don't think this enables anything fancy

> Do we assume a constant filled work pipe?

Yes, latency is not a concern (at least right now, but expect to have deadlines given by FSM at some point in the future)

> Are we going to support work stealing?

I thought about this, but this shouldn't be needed much assuming a decent work backlog and small local queues

> what does it win us?

Well, vs the current code, 50% faster precommit2, and 10-20% faster precommit1 when using remote workers.

> what is the cost of moving the next stage to another node vs doing it on the same node?

It's just network transfer between ingest/compute nodes (+storage array on the last step) + a bunch of sequential IO on both ends. Currently we need to do this just because of how different hardware profiles precommit 1 / 2 requires, but when we get 10x faster sealing, just moving 32G sectors will have big impact

**Comment by icorderi** [comment]
Is the cost of fetching the next item too high compared to cost of running it?
> Yes, sector cache for precommit2 is in the 350-400G range last time I checked, and moving that over 1Gbps (which if we can we should support as best as we can just based on how much cheaper it is) takes ~1h.

> Precommit2 compute takes 51min on a 32 core AMD Threadripper, 

That for me is a no. The cost of deciding to do a precommit2 is minimal (network roundtrip + pop/lookup) compared to the transfer of 350GB + 50min of crunching. 
This is looking more to me as keep all the work items in the same place so we can make real time decisions on what to schedule where. (with some redundancy for availability reasons, I think actually query cost for next item does not need to scale beyond a node right now, we can figure out a partition scheme later) 

But I see we have a difference on what we consider the contents of the "work queue" vs the work itself. 

> which means that most of the time is spend just waiting. 

If we consider the the data transfer part of the contents of the work queue as you propose, then we have a situation where the arrival rate is 1h and the service time in 50min. 

Simplifying this to an M/M/1.
The steady state of that is a service with 50% utilization.  
You are not going to change that by deepening the work queue or starting with a larger buffer. 

If you want that node to go above 50% utilization we need to double up on the network side (fatter pipe or more pipes).

After that, it's just a matter of keeping the next sector in local storage while the current crunching is in process. Which I guess it what you call having a small queue of 2-3 items. 
In practice you only need a queue depth of 1 with those service times. 
Depending on the IO needs, that waiting item can be kept on specialized local storage.   
So if the service occurs directly form storage (I'm assuming this 300GB are not entirely in ram) then the storage needs are 2x. 

> And as sealing becomes faster, this will become an even bigger issue.
I'm not sure what is a bigger problem, the service being less utilized or the queue depth skyrocketing when that service time starts getting really small. 
There is a limit on the data transfer, maintaining saturated data pipes of sectors coming in will quickly start eating into the CPU. This will in-turn increase the service time. 
Whatever that new steady state depth is, can cause issues depending on the hardware requirements for the specialized storage and bus bandwidth on the node.

--- 

I would consider the item in the work queue as "precommit2 sector8272" and that being what is scheduled to node A (That work implying data transfer + crunching). 

If we want a bit more flexibility, then we can split the task into: 
- "transfer sector8272 to staging 762"
- "precommit2 sector8272" 

This allows multiple nodes to be able to serve "precommit2 sector8272", by sharing the storage called "staging 762". 
It is likely any node at same distance from the current location of "sector8272" can make the transfers to "staging 762" (We can always add heuristics for doing dynamic balancing, to select a better option). 

---

what does it win us?
> Well, vs the current code, 50% faster precommit2, and 10-20% faster precommit1 when using remote workers.

I don't think that math is correct, without changing the arrival rate, you are still at 50% utilization. The reasons I gave before, but we might just be discussing semantics at this point.

**Comment by magik6k** [comment]
I like 'transfer' as a separate thing we can tell the worker to do, it may even fit nicely into retrieval data-transfer nodes (nodes bridging internet, retrieval market, and storage system)

Ultimately what we want to achieve here is to be able to fetch data without locking resources in the scheduler.

I'll try to scribble something when I'm freed from the windowed post work, I feel like this can be done in <100 lines of code, most of which will be implementing the 'transfer x to y' logic, worst case we'll have something that works and can be easily changed.

**Comment by icorderi** [comment]
Yes, separating the data fetch from the crunching and having a node being able to do both in parallel is good improvement.

**Comment by magik6k** [comment]
(done in #20)
